src/rdkafka_ssl.c

```diff
@@ -999,6 +1000,56 @@ static int rd_kafka_ssl_set_certs (rd_kafka_t *rk, SSL_CTX *ctx,
                 check_pkey = rd_true;
         }
 
+#if OPENSSL_VERSION_NUMBER >= 0x10100000
+        if (rk->rk_conf.ssl.ptr_engine) {
+                STACK_OF(X509_NAME)* cert_names = sk_X509_NAME_new_null();
+                STACK_OF(X509_OBJECT)* roots = X509_STORE_get0_objects(SSL_CTX_get_cert_store(ctx));
+                X509* x509 = NULL;
+                EVP_PKEY* pkey = NULL;
+                int i = 0;
+                for (i = 0; i < sk_X509_OBJECT_num(roots); i++) {
+                    x509 = X509_OBJECT_get0_X509(sk_X509_OBJECT_value(roots, i));
+                    if (x509)
+                        sk_X509_NAME_push(cert_names, X509_get_subject_name(x509));
+                }
+
+                x509 = NULL;
+                r = ENGINE_load_ssl_client_cert(rk->rk_conf.ssl.ptr_engine, NULL,
+                    cert_names,
+                    &x509,
+                    &pkey,
+                    NULL, NULL, NULL);
+
+                if (cert_names)
+                        sk_X509_NAME_free(cert_names);
+
+                if (r == -1 || !x509 || !pkey) {
+                        rd_snprintf(errstr, errstr_size,
+                                    "ENGINE_load_ssl_client_cert failed");
```

Comment:
Since the SSL error stack error message is automatically appended these strings need to end with `"..failed: "`

Reply:
sure will change it in next iteration.